### PR TITLE
Defined pre-conditions and error messages for the marconi-sidechain JSON-RPC API specification

### PR DESCRIPTION
**Pull request by koslambrou**
* Did some rewrite in the API.adoc in order to reduce the JSON-Schema duplication across JSON-RPC methods
* Documented the pre-conditions and JSON-RPC error object for each endpoint. Basically, what are the failing scenarios when calling an endpoint?

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: marconi-sidechain/doc/API.adoc [hunk 1]
```diff
@@ -147,745 +112,585 @@ Taking the chainpoint before ensure that we have consistent infomation across al
...
+      "const": "The 'policyId' param value must be a valid minting policy hash."
+    },
+    {
+      "const": "The 'afterTx' param value must be a valid transaction ID."
+    },
+    {
+      "const": "The 'policyId' and 'assetName' param values must belong in the provided target 'AssetId's."
+    },
+    {
+      "const": "The 'afterTx' param value must be an existing transaction ID in the Cardano network."
```
Comment (berewt) [inline marconi-sidechain/doc/API.adoc:500]: Change message (see the previous comment about _tracked_ transactions.)

File: marconi-sidechain/doc/API.adoc [hunk 2]
```diff
@@ -147,745 +112,585 @@ Taking the chainpoint before ensure that we have consistent infomation across al
...
 
+
+=== getActiveStakePoolDelegationByEpoch
+
+Retrieves the active stake pool delegation per epoch.
+
 ==== Pre-conditions
 
-TBD
+* The `epochNo` param value must be a natural number.
```
Comment (berewt) [inline marconi-sidechain/doc/API.adoc:603]: I don't remember, do we track it for all `epochNo`? If no, shouldn't this restriction be clearly stated?

**Review by berewt — COMMENTED** [review]
A few comments and additions. I didn't take a close look at `getNonceByEpoch` and `getActiveStakePoolDelegationByEpoch` as I'm not that familiar with them and they don't look totally finished yet. Do you want to document them before closing this PR or would it be done in another PR?

File: marconi-sidechain/doc/API.adoc [hunk 3]
```diff
@@ -147,745 +112,585 @@ Taking the chainpoint before ensure that we have consistent infomation across al
...
+[%collapsible]
+====
+```JSON
+{
+  "oneOf": [
+    {
+      "const": "The 'address' param value must be in the Bech32 format."
+    },
+    {
+      "const": "The 'address' param value must belong in the provided target addresses."
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc:356]: Why "to"? The use of "belong in" basically means "belons inside" which what I meant?

File: marconi-sidechain/doc/API.adoc (hunk 2 above)
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc:603]: The other error message below basically says that: "The 'epochNo' param value must belong in an era that has nonces (Byron era doesn't have nonces)."

File: marconi-sidechain/doc/API.adoc [hunk 4]
```diff
@@ -147,745 +112,585 @@ Taking the chainpoint before ensure that we have consistent infomation across al
...
+{}
+```
 ====
 
-==== Rationale
-
 [NOTE]
 ====
-In here, we'll document the rationale for having this method => why does the sidechain application need it.
+Whatever you put in the param object, this method will just ignore it.
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc]: Changed it to what we said. See here https://github.com/input-output-hk/marconi/pull/56/files#diff-a894fbddaf704d7d5d99ab43c2fcdfb9054706009c6f0b16d0d82b0ff5f89564R148

File: marconi-sidechain/doc/API.adoc [hunk 5]
```diff
@@ -147,745 +112,585 @@ Taking the chainpoint before ensure that we have consistent infomation across al
...
+
+* The `address` param value must be in the Bech32 format.
+
+* If the target addresses used for `marconi-sidechain` is non empty, then the `address` param value should be part of the target addresses.
+
+* The `createdAfterSlotNo` and `unspentBeforeSlotNo` param values must be natural numbers starting at 0.
+
+* The `unspentBeforeSlotNo` param value should be larger than the `createdAfterSlotNo`.
+
+* The `createdAfterSlotNo` param value must be lower than the current indexed `SlotNo` (returned by the `getCurrentSyncedBlock` endpoint).
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc]: Yep it is. Changed it.

File: marconi-sidechain/doc/API.adoc [hunk 6]
```diff
@@ -147,745 +112,585 @@ Taking the chainpoint before ensure that we have consistent infomation across al
...
+
+Retrieves all token burning events given an `AssetId` that have occurred before a given slot, and optionally after a given transaction ID.
+
 ==== Pre-conditions
 
-TBD
+* The `policyId` param value must be a valid minting policy hash.
 
-==== Post-conditions
+* The `afterTx` param value must be a valid transaction ID.
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc:401]: I kept this pre-conditions as I'm referring to the encoding of the transaction ID.

However, I added another pre-condition about the fact that transaction ID but exist in the network and it must have burned a token: https://github.com/input-output-hk/marconi/pull/56/files#diff-a894fbddaf704d7d5d99ab43c2fcdfb9054706009c6f0b16d0d82b0ff5f89564R405

File: marconi-sidechain/doc/API.adoc [hunk 7]
```diff
@@ -147,745 +112,585 @@ Taking the chainpoint before ensure that we have consistent infomation across al
...
 
-TBD
+* If the list of target `AssetId` used for `marconi-sidechain` is non empty, then the `(policyId, assetName)` should be part of that list.
 
-==== Method-specific errors
+* The `afterTx` param value must be an existing transaction ID in the Cardano network.
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc]: Yep, changed it to: https://github.com/input-output-hk/marconi/pull/56/files#diff-a894fbddaf704d7d5d99ab43c2fcdfb9054706009c6f0b16d0d82b0ff5f89564R405

File: marconi-sidechain/doc/API.adoc (hunk 2 above)
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc:630]: Good point. Added it as follows:

```
The `nonce` result value must be in an era that has epoch nonces (for example, Byron era doesn't have nonces).
```

File: marconi-sidechain/doc/API.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-sidechain/doc/API.adoc]: I always thought "belong in" meant "it's should be in it but it's not necessary in" but I'm not a native speaker, so I can be totally wrong, I trust you then. :)

[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-sidechain/doc/API.adoc]: Ah good point :D Not a native speaker as well.

**Review by berewt — APPROVED** [review]
We can ship it! 🚀

**Pull request merged**